net/inet_ntop.c

```diff
@@ -1,27 +1,115 @@
+/*
+ * Phoenix-RTOS
+ *
+ * libphoenix
+ *
+ * net/inet_ntop.c
+ *
+ * Copyright 2021 Phoenix Systems
+ * Author: Maciej Purski
+ *
+ * This file is part of Phoenix-RTOS.
+ *
+ * %LICENSE%
+ */
+
 #include <arpa/inet.h>
 #include <string.h>
 #include <stdio.h>
+#include <errno.h>
+
 
-const char *inet_ntop(int af, const void *restrict src,
-		      char *restrict dst, socklen_t size)
+static const char *inet_ntop4(const void *src, char *dst, socklen_t size)
 {
-	if (af == AF_INET) {
-		if (size < 16 || dst == NULL || src == NULL)
-			return NULL;
+	if (size < INET_ADDRSTRLEN) {
+		errno = ENOSPC;
+		return NULL;
+	}
 
-		snprintf(dst, size, "%u.%u.%u.%u",
-				(unsigned int)((unsigned char *)src)[0],
-				(unsigned int)((unsigned char *)src)[1],
-				(unsigned int)((unsigned char *)src)[2],
-				(unsigned int)((unsigned char *)src)[3]);
+	snprintf(dst, size, "%u.%u.%u.%u",
+			 (unsigned int)((unsigned char *)src)[0],
+			 (unsigned int)((unsigned char *)src)[1],
+			 (unsigned int)((unsigned char *)src)[2],
+			 (unsigned int)((unsigned char *)src)[3]);
 
-		return dst;
-	} else if (af == AF_INET6) {
-		memset(dst, 0, size);
-		memcpy(dst, "::", strlen("::"));
+	return dst;
+}
 
+static const char *inet_ntop6(const void *src, char *dst, socklen_t size)
+{
+	const struct in6_addr *addr;
+	int i, zero_idx = -1, tmp_len, zero_len = 0, tmp_idx = -1;
+	char *dst_ptr = dst;
+	const int n_hextets = sizeof(addr->s6_addr16) / sizeof(addr->s6_addr16[0]);
+
+	if (size < INET6_ADDRSTRLEN) {
+		errno = ENOSPC;
+		return NULL;
+	}
+	memset(dst, 0, size);
+
+	addr = (const struct in6_addr *)src;
+	if (IN6_IS_ADDR_V4MAPPED(addr) || IN6_IS_ADDR_V4COMPAT(addr)) {
+		snprintf(dst, INET6_ADDRSTRLEN, "::%s%u.%u.%u.%u",
+				 addr->s6_addr16[5] == 0xFFFF ? "ffff:" : "",
+				 addr->s6_addr[12],
+				 addr->s6_addr[13],
+				 addr->s6_addr[14],
+				 addr->s6_addr[15]);
 		return dst;
 	}
 
+	/* Find the longest all-zero string */
+	for (i = 0; i < n_hextets; i++) {
+		if (addr->s6_addr16[i] == 0) {
+			if (tmp_idx == -1) {
+				/* Found new all-zero string */
+				tmp_idx = i;
+				tmp_len = 1;
+			}
+			else {
+				tmp_len++;
+			}
+		}
+		/* Check if, all-zero string is over */
+		if (tmp_idx != -1 && (addr->s6_addr16[i] != 0 || i == n_hextets - 1)) {
+			if (tmp_len > 1 && tmp_len > zero_len) {
+				zero_len = tmp_len;
+				zero_idx = tmp_idx;
+			}
+			tmp_idx = -1;
+		}
+	}
+
+	for (i = 0; i < n_hextets;) {
+		if (i == zero_idx) {
+			if (i == 0)
+				*(dst_ptr++) = ':';
+			*(dst_ptr++) = ':';
+			i += zero_len;
+		} else {
+			dst_ptr += snprintf(dst_ptr, 5, "%x", ntohs(addr->s6_addr16[i]));
+			if (i != n_hextets - 1)
+				*(dst_ptr++) = ':';
+			i++;
+		}
```

Comment:
I don't know how the IPv6 addresses are usually printed but it seems that common parts could be moved to outside of the if-else block?
```suggestion
		if (i == zero_idx) {
			if (i == 0)
				*(dst_ptr++) = ':';
			i += zero_len;
		} else {
			dst_ptr += snprintf(dst_ptr, 5, "%x", ntohs(addr->s6_addr16[i]));
			i++;
		}

		if (i != n_hextets)
				*(dst_ptr++) = ':';
```

Reply:
Unfortunately, your suggestion does not work properly, when in a case, where the address ends with an all-zero string. For example, after applying your change, the function would produce: `fd00:0:8000:80:` instead of `fd00:0:8000:80::`.
Maybe, I'll think it over and find a better way to simplify this ugly loop.